WeatherApp/Model/WeatherManager.swift

```diff
@@ -10,95 +10,55 @@ import Foundation
-        
+
         return removedDiacritics
     }
-    
-    func performWeatherRequest(with url: String, fromLocation: Bool) {
-        Alamofire.request(url).response { (responseData) in
-            if let weatherData = responseData.data {
-                guard var weatherModel = self.parseJSON(with: weatherData) else {
-                    return
-                }
-                
-                weatherModel.fromLocation = fromLocation
-                self.delegate?.weatherDataDidUpdate(self, weather: weatherModel)
+
+    private func performWeatherRequest(with url: String, fromLocation: Bool) {
+        Alamofire.request(url).response { (response) in
+            guard let data = response.data else {
+                return
             }
-        }
-    }
-    
-    func parseJSON(with weatherData: Data) -> WeatherModel? {
-        let data = JSON(weatherData)
-        
-        if data["cod"].stringValue != "200" {
-            return nil
-        }
-        
-        let cityNameValue = data["city"]["name"].stringValue
-        
-        var dayForecast: [WeatherModel.DayForecast] = []
-        
-        data["list"].arrayValue.forEach { (day) in
-            let temp = day["main"]["temp"].intValue
-            let condId = day["weather"][0]["id"].intValue
-            let description = day["weather"][0]["description"].stringValue
-            let dateTime = day["dt_txt"].stringValue.split(separator: " ")
-            let date = reformatDate(in: String(dateTime[0]))
-            let time = reformatTime(in: String(dateTime[1]))
-            
-            dayForecast.append(WeatherModel.DayForecast(conditionID: condId, temp: temp, description: description, date: date, time: time))
-        }
-        
-        return WeatherModel(cityName: cityNameValue, dayForecast: dayForecast, fromLocation: false)
-    }
-    
-    func reformatDate(in value: String) -> String {
-        let dateFormatterGet = DateFormatter()
-        dateFormatterGet.dateFormat = "yyyy-MM-dd"
 
-        let dateFormatter = DateFormatter()
-        dateFormatter.dateFormat = "E, d MMM"
+            guard let model = self.mapDataToModel(data, fromLocation: fromLocation) else {
+                return
+            }
 
-        let date: Date? = dateFormatterGet.date(from: value)
-        
-        return dateFormatter.string(from: date!)
+            self.delegate?.weatherDataDidUpdate(self, weather: model)
+        }
     }
-    
-    func reformatTime(in value: String) -> String {
-        let dateFormatterGet = DateFormatter()
-        dateFormatterGet.dateFormat = "HH:mm:ss"
 
-        let dateFormatter = DateFormatter()
-        dateFormatter.dateFormat = "HH:mm"
+    private func mapDataToModel(_ data: Data, fromLocation: Bool) -> WeatherModel? {
+        guard let responseModel = try? JSONDecoder().decode(WeatherResponse.self, from: data) else {
```

Comment:
Dodałem dodatkową warstę mapowania dla `WeatherModel` w postaci `WeatherResponse`. Dzięki temu mamy odseparowane dane z API od danych używanych do logiki aplikacji lub budowania widoków.

Reply:
Nie potrzebujemy tutaj `SwiftyJson`. iOS dostarcza nam narzędzia do mapowania `JSON` na obiekty.
`JSONDecoder().decode(WeatherResponse.self, from: data)` zwraca instancję obiektu typu `WeatherResponse`, który tworzy z `data: Data`. `WeatherResponse` musi implementować protokół `Decodable`

Reply:
`try?` jest po to, bo metoda `decode(...)` może wyrzucić wyjątek. Pytajnik działa tak, że jeśli cokolwiek pójdzie nie tak to `decode(...)` zwróci `nil`, zamiast rzucać wyjątkiem. Pomijamy po prostu obsługę wyjątku za pomocą `do {} catch {}`